### PR TITLE
Speed up tile building: don't copy GraphTileBuilder every time

### PR DESCRIPTION
**Pull request by merkispavel**
### What issue is this PR targeting? If there is no issue that addresses the problem, please open a corresponding issue and link it here.

Profiling `valhalla_build_tiles` target on my machine(mac OS) showed that `GraphTileBuilder` copy-construction and destruction take significant amount of time. Here is stacktrace from OS X "Time Profiler" 
Screen Shot 2020-07-13 at 22 06 12

The further analysis brought me to the `IsNextEdgeInternal` function. As you see the first line of the function copies `GraphTileBuilder`. https://github.com/valhalla/valhalla/blob/2ac1fa009c305e6f472c032e89f938e59c25cbe1/src/mjolnir/graphenhancer.cc#L884

The next experiments showed 10+% performance improvements:
| Source osm pbf  | pbf size | time without fix | time with fix | delta % |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| belarus | 210 Mb  | 4:18 | 3:32 | 17.8% |
| germany | 3.4 Gb | ~1:36:00 | ~1:21:00 | 15.6% |

UPD:
the default valhalla config was used for experiments

Open questions:
1) Do we have regression tests to make sure that I didn't mess up the tile structure?
2) Do we need to benchmark the fix on larger amount of data? My expectations(hopes) that profit will be ~linear to the data size scale. However, It may depend on the number of threads/other specs of the server machine/config where the whole world tiles are built.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/mjolnir/graphenhancer.cc [hunk 1]
```diff
@@ -874,61 +874,69 @@ void GetHeadings(GraphTileBuilder& tile, NodeInfo& nodeinfo, uint32_t ntrans) {
...
 }
 
-// Is the next edge from the end node of the directededge is internal or not.
-bool IsNextEdgeInternal(const DirectedEdge directededge,
-                        GraphTileBuilder& tilebuilder,
-                        GraphReader& reader,
-                        std::mutex& lock,
-                        bool infer_internal_intersections) {
-  // Get the tile at the startnode
-  GraphTileBuilder tile = tilebuilder;
```
Comment (merkispavel) [inline src/mjolnir/graphenhancer.cc:884]: Looks like a redundant copy

**Comment by gknisely** [comment]
> 1.  Do we have regression tests to make sure that I didn't mess up the tile structure?
> 2.  Do we need to benchmark the fix on larger amount of data? My expectations(hopes) that profit will be ~linear to the data size scale. However, It may depend on the number of threads/other specs of the server machine/config where the whole world tiles are built.

We should test this out in the test environment and RAD test it.

**Comment by kevinkreiser** [comment]
now that tile builds are stable, you can do a before and after checksum of the tiles created with utrecht for example to prove that this is safe, without RAD imho

**Comment by merkispavel** [comment]
> now that tile builds are stable, you can do a before and after checksum of the tiles created with utrecht for example to prove that this is safe, without RAD imho. maybe just in case use a larger pbf like switzerland or pennsylvania or some bigger geographical area.

oh, god lord. Tried to compare utrecht tiles hashsums - different. Did tile-building twice with master - level 2 tile has flaky hashsum :( Looking for the cause at the moment.

UPD: looks like edge bins are different. Guess: maybe edges are stored in different order

**Comment by kevinkreiser** [comment]
@merkispavel yeah i can see why that is happening. we do a random shuffle on the tilequeue so different threads get different tiles to bin. there are two things that get added to the bins, the edges that originate in the tile, i think those are sorted so should be a problem and then edges that cross through the tile but dont originate there (tweeners). those results come from different threads and are added on to the bins at the end. since the tiles are shuffled randomly this would cause different tweeners to come from different threads but we iterate through the threads in a fixed order at the end. this means the tweeners are in a randomized order which means we can do 1 of 2 things. pick a constant seed for the random shuffle, or sort the tweeners before we append them to the bins. the seed is probably the easiest but it might be nice (although dubiously useful) to have a more "sensible" ordering of them.

**Comment by merkispavel** [comment]
> @merkispavel yeah i can see why that is happening. we do a random shuffle on the tilequeue so different threads get different tiles to bin. there are two things that get added to the bins, the edges that originate in the tile, i think those are sorted so shouldn't be a problem and then edges that cross through the tile but dont originate there (tweeners). those results come from different threads and are added on to the bins at the end. since the tiles are shuffled randomly this would cause different tweeners to come from different threads but we iterate through the threads in a fixed order at the end. this means the tweeners are in a randomized order which means we can do 1 of 2 things. pick a constant seed for the random shuffle, or sort the tweeners before we append them to the bins. the seed is probably the easiest but it might be nice (although dubiously useful) to have a more "sensible" ordering of them.

Turned out that utrecht_tiles are built with 1 thread(hardcoded in cmake) so seed wasn't a problem.
Fix comes with https://github.com/valhalla/valhalla/pull/2515

**Comment by merkispavel** [comment]
I've tested these changes with https://github.com/valhalla/valhalla/pull/2515: utrecht tiles checksums do match.
I guess the last question here is whether tests should be added to increase code coverage

**Comment by kevinkreiser** [comment]
@merkispavel i think we always want to be increasing test coverage, but since you didnt actually add anything new here i dont think adding coverage should really be done in this pr. to me we should look at the report and focus on the places with least coverage over time. i wouldnt block this pr over the coverage check and can merge anyway when we are ready to go

**Comment by kevinkreiser** [comment]
@gknisely pavel ran this on a larger tile set (258 tiles in PA) with parallelism turned on and confirmed that the output is the same (checksums of tiles in directories match with master)

**Pull request merged**